### PR TITLE
MoE - Token dropping for Full Tensor Paralellism

### PR DESCRIPTION
**Pull request by siddharth9820**
This PR enables token dropping for full tensor parallelism. 
Also corrects timers.

(Still WIP)

### CODE REVIEW

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 1]
```diff
@@ -530,6 +530,7 @@ def _configure_moe_settings(self):
             logger.warn(
                 "ZeRO Stage 1 has not been thoroughly tested with MoE. This configuration is still experimental."
             )
+            self.contiguous_gradients = True
```
[the remark this answers is not in the document]
Comment (siddharth9820) [reply deepspeed/runtime/zero/stage_1_and_2.py:533]: the engine sets this to False

File: deepspeed/runtime/engine.py [hunk 2]
```diff
@@ -176,6 +177,26 @@ def __init__(self, enable_micro_timers, enable_global_timers):
...
+
+    def get_profile(self):
+        new_counter_data = np.array([self.read(i) for i in range(self.total)])
+        data = (new_counter_data - self.current_counter_data) * 4 / 1024 / 1024
+        self.current_counter_data = new_counter_data
+        return data
+
+    def read(self, port_id, counter_name="port_rcv_data"):
+        file_name = f"/sys/class/infiniband/mlx5_ib{port_id}/ports/1/counters/{counter_name}"
+        with open(file_name, "r") as f:
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:195]: Will this fail on systems without inifiband?

File: deepspeed/runtime/engine.py [hunk 3]
```diff
@@ -176,6 +177,26 @@ def __init__(self, enable_micro_timers, enable_global_timers):
             ]
 
 
+class InfinibandNetworkCounters(object):
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:180]: Can we make this a separate PR?

**Review by tjruwase — APPROVED** [review]

**Comment by jomayeri** [comment]
@siddharth9820 Can you please respond if you want to keep this PR open, otherwise I will close it. Thanks.
